### PR TITLE
Discord authentication.

### LINKED ISSUE #157: Discord Provider
I would like to request the addition of Discord if anyone is up to it =D 

 https://discordapp.com/developers/docs/topics/oauth2

### PR DESCRIPTION
**Pull request by dialupnoises**
I implemented Discord authentication using the Discord OAuth2 API: https://discordapp.com/developers/docs/topics/oauth2

The code is based on the Reddit provider, and by "based" I mean "mostly copied and pasted."

I didn't realize until after I finished it that someone had requested it. In that case, this would close #157.

### CODE REVIEW

**Comment by TerribleDev** [comment]
Thank you very much for the contribution!! I'll look this over wednesday-ish and get back to you.

File: src/Owin.Security.Providers.Discord/DiscordAuthenticationHandler.cs [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+        {
+            _httpClient = httpClient;
+            _logger = logger;
+        }
+
+        protected override async Task<AuthenticationTicket> AuthenticateCoreAsync()
+        {
+            AuthenticationProperties properties = null;
+
+#if !DEBUG
```
Comment (TerribleDev) [inline src/Owin.Security.Providers.Discord/DiscordAuthenticationHandler.cs]: can you remove the compiler directive? if you are looking to catch the error while debugging there are numerous things you can invoke System.Diagnostics.Debugger.Break() inside of an if System.Diagnostics.Debugger.IsAttached  == true statement, or even tell VS to throw even on user handled exceptions.

File: src/Owin.Security.Providers.Discord/DiscordAuthenticationHandler.cs [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+                if (!string.IsNullOrEmpty(context.Id))
+                {
+                    context.Identity.AddClaim(new Claim(ClaimTypes.NameIdentifier, context.UserName, XmlSchemaString, Options.AuthenticationType));
+                }
+                if (!string.IsNullOrEmpty(context.UserName))
+                {
+                    context.Identity.AddClaim(new Claim(ClaimsIdentity.DefaultNameClaimType, context.UserName, XmlSchemaString, Options.AuthenticationType));
+                }
+				if(!string.IsNullOrEmpty(context.Email))
+				{
```
Comment (TerribleDev) [inline src/Owin.Security.Providers.Discord/DiscordAuthenticationHandler.cs]: can you fix the indentations here?

File: src/Owin.Security.Providers.Discord/Provider/DiscordAuthenticatedContext.cs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+            User = user;
+            AccessToken = accessToken;
+            RefreshToken = refreshToken;
+            int expiresValue;
+            if (int.TryParse(expires, NumberStyles.Integer, CultureInfo.InvariantCulture, out expiresValue))
+            {
+                ExpiresIn = TimeSpan.FromSeconds(expiresValue);
+            }
+			Id = TryGetValue(user, "id");
+			UserName = TryGetValue(user, "username");
```
Comment (TerribleDev) [inline src/Owin.Security.Providers.Discord/Provider/DiscordAuthenticatedContext.cs]: can you fix the indentations here as well?

**Comment by dialupnoises** [comment]
I keep forgetting to fix my indentation before I make commits. I had meant to remove that compiler directive, but I guess I'd forgotten about that too!

**Comment by TerribleDev** [comment]
great thanks it looks good...when I get home, and have time I will merge and publish this...thanks!

**Pull request merged**